### PR TITLE
Add user status

### PR DESCRIPTION
**Pull request by swathikrishna-kv**

### CODE REVIEW

File: src/schema/graphql.schema.ts [hunk 1]
```diff
@@ -240,6 +240,7 @@ export interface User {
     firstName: string;
     middleName?: string;
     lastName: string;
+    status: string;
```
Comment (doylefermi-kv) [inline src/schema/graphql.schema.ts:243]: Can we make use of [Graphql Enum](https://graphql.org/learn/schema/#enumeration-types) to ensure this is a well defined value and not some random string?

File: src/authorization/entity/user.entity.ts [hunk 2]
```diff
@@ -38,6 +38,12 @@ class User extends BaseEntity {
 
   @Column({ nullable: true, default: false })
   public twoFAEnabled?: boolean;
+
+  @Column({ default: 'inactive' })
+  public status!: string;
```
Comment (doylefermi-kv) [inline src/authorization/entity/user.entity.ts:43]: Can we make use of the [Typeorm Enum](https://github.com/typeorm/typeorm/blob/master/docs/entities.md#enum-column-type) datatype here?

File: src/authorization/entity/user.entity.ts [hunk 3]
```diff
@@ -38,6 +38,12 @@ class User extends BaseEntity {
...
+
+  @Column({ nullable: true })
+  public inviteToken?: string;
```
Comment (doylefermi-kv) [inline src/authorization/entity/user.entity.ts:46]: Is this part of this PR? Or a separate feature?

File: src/migrations/1666763401586-User Status Invite Token.ts [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+export class UserStatusInviteToken1666763401586 implements MigrationInterface {
+  name = 'UserStatusInviteToken1666763401586';
+
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(`DROP INDEX "user_phone_unique_idx"`);
+    await queryRunner.query(
+      `ALTER TABLE "user" ADD "status" character varying NOT NULL DEFAULT 'inactive'`,
+    );
+    await queryRunner.query(
+      `ALTER TABLE "user" ADD "invite_token" character varying`,
```
Comment (doylefermi-kv) [inline src/migrations/1666763401586-User Status Invite Token.ts:12]: Is this part of this PR or a different feature?

File: src/migrations/1666763401586-User Status Invite Token.ts [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    );
+    await queryRunner.query(
+      `ALTER TABLE "entity_permission" DROP CONSTRAINT "PK_ac18341de235a9dddb822c35892"`,
```
Comment (doylefermi-kv) [inline src/migrations/1666763401586-User Status Invite Token.ts:15]: Why are we dropping this constraint?

File: src/migrations/1666763401586-User Status Invite Token.ts [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+      `ALTER TABLE "entity_permission" ADD CONSTRAINT "PK_22d409e099ab8a6bc3fc7b7b8a1" PRIMARY KEY ("permission_id", "entity_id")`,
+    );
+    await queryRunner.query(
+      `ALTER TABLE "entity_permission" DROP CONSTRAINT "PK_22d409e099ab8a6bc3fc7b7b8a1"`,
+    );
+    await queryRunner.query(
+      `ALTER TABLE "entity_permission" ADD CONSTRAINT "PK_451db0ef409dd0b566c4efca6a3" PRIMARY KEY ("entity_id")`,
+    );
+    await queryRunner.query(
+      `ALTER TABLE "entity_permission" DROP COLUMN "permission_id"`,
```
Comment (doylefermi-kv) [inline src/migrations/1666763401586-User Status Invite Token.ts:27]: Please review this change.

File: src/authorization/service/user.service.ts [hunk 7]
```diff
@@ -190,6 +193,10 @@ export default class UserService {
     if (!user) {
       throw new UserNotFoundException(id);
     }
+    if (user.status == 'inactive') {
```
Comment (doylefermi-kv) [inline src/authorization/service/user.service.ts:196]: Can we reuse the Enum datatype here for status?

File: src/authorization/service/user.service.ts [hunk 8]
```diff
@@ -190,6 +193,10 @@ export default class UserService {
...
+      throw new AccountIsInactive();
```
Comment (doylefermi-kv) [inline src/authorization/service/user.service.ts:197]: Can we rename this to `InactiveAccountException()`?

File: src/authorization/service/user.service.ts [hunk 9]
```diff
@@ -190,6 +193,10 @@ export default class UserService {
...
+    }
+    await this.updateField(id, 'status', 'inactive');
```
Comment (doylefermi-kv) [inline src/authorization/service/user.service.ts:199]: What if softDelete fails? Shouldn't we make this operation atomic?

File: src/authorization/exception/user.exception.ts [hunk 10]
```diff
@@ -1,7 +1,14 @@
...
 
 export class UserNotFoundException extends NotFoundException {
   constructor(userId: string) {
     super(`User ${userId} not found`);
   }
 }
+
+export class AccountIsInactive extends BadRequestException {
+  constructor() {
+    super(`Account is inactive`);
```
Comment (doylefermi-kv) [inline src/authorization/exception/user.exception.ts:12]: What would be the status code thrown for this exception? `BadRequestException` throws `400` which indicates that the request is incorrect. Can we put in a more appropriate status code here?

File: src/authorization/graphql/user.graphql [hunk 11]
```diff
@@ -5,6 +5,7 @@ type User {
   firstName: String!
   middleName: String
   lastName: String!
+  status: String!
```
Comment (doylefermi-kv) [inline src/authorization/graphql/user.graphql:8]: Can we make use of [Graphql Enum](https://graphql.org/learn/schema/#enumeration-types) to ensure this is a well defined value and not some random string?

File: src/authorization/exception/user.exception.ts [hunk 12]
```diff
@@ -1,7 +1,14 @@
+import { BadRequestException } from '@nestjs/common';
```
Comment (doylefermi-kv) [inline src/authorization/exception/user.exception.ts:1]: Should this exception be thrown globally whenever an inactive user tries to perform operations?

**Review by doylefermi-kv — CHANGES_REQUESTED** [review]

**Review by doylefermi-kv — CHANGES_REQUESTED** [review]

File: src/authorization/service/user.service.ts [hunk 13]
```diff
@@ -5,11 +5,15 @@ import { Connection, Repository } from 'typeorm';
...
 import {
   OperationType,
+  Status,
   UpdateUserGroupInput,
   UpdateUserInput,
   UpdateUserPermissionInput,
 } from '../../schema/graphql.schema';
-import { UserNotFoundException } from '../exception/user.exception';
+import {
+  InactiveAccountException,
```
Comment (doylefermi-kv) [inline src/authorization/service/user.service.ts]: Can we remove this unused import?

File: src/authorization/exception/user.exception.ts [hunk 14]
```diff
@@ -1,3 +1,4 @@
+import { ForbiddenException } from '@nestjs/common';
```
Comment (doylefermi-kv) [inline src/authorization/exception/user.exception.ts:1]: Please remove this unused import

Comment (doylefermi-kv) [inline src/authorization/exception/user.exception.ts:1]: Can we remove this unused import?

**Review by doylefermi-kv — APPROVED** [review]

**Review by JoyalAJohney-KeyValue — APPROVED** [review]

**Pull request merged**